### DISCUSSION TITLE
Configure let's encrypt certificate to work with emqx

### DISCUSSION THREAD

**Discussion by FernandoGarcia** [emqx/emqx#10762] [Q&A] [answered]
Hi!

I have this file types available on my server:

```
-rw------- 1 root root 1883 May 20 23:04 cert4.pem
-rw------- 1 root root 3749 May 20 23:04 chain4.pem
-rw-r--r-- 1 root root 5632 May 20 23:04 fullchain4.pem
-rw------- 1 root root 1704 May 20 23:04 privkey4.pem
```

I would like to know what's the correct path to place in `emqx.conf`.

I have tried many combinations without success.

In the last attempt I have used in this way:

```
  ssl_options.keyfile = "/etc/letsencrypt/live/domain.com/privkey4.pem"
  ssl_options.certfile = "/etc/letsencrypt/live/domain.com/cert4.pem"
  ssl_options.cacertfile = "/etc/letsencrypt/live/domain.com/fullchain4.pem"
```

I'm getting this type of error:

```
May 20 23:45:14 domain.com bash[10292]: 2023-05-20T20:45:14.624566-03:00 [error]     msg: cert_file_not_found
May 20 23:45:14 domain.com bash[10292]:     mfa: emqx_tls_lib:resolve_cert_path_for_read_strict/1
May 20 23:45:14 domain.com bash[10292]:     line: 584
May 20 23:45:14 domain.com bash[10292]:     path: /etc/letsencrypt/live/domain.com/fullchain.pem
```

I have emqx v5.0.25 running on Ubuntu 22.04.

Thanks in advance.

**Comment by lafirest** [comment]
@FernandoGarcia 
Hi, where is the `fullchain.pem`, I only see a file named `fullchain4.pem`.

**Comment by FernandoGarcia** [comment]
Hi!

Thanks for your answer.

I just forgot to edit the file name in error message because in fact in live/domain.com/ there's symlinks pointing to /archive/domain.com/. 

In archive there's a 4 on name but in live not. 

I have showed the permission from origin folder.

EDIT:

Here you can see the real path.

![Captura de tela de 2023-05-24 11 27 42](https://github.com/emqx/emqx/assets/2778501/869820c6-fbae-4369-af3d-7acd2813d761)

My question is about the correct file type because I don't have a `cacert.pem` file. The options are `chain.pem` or `fullchain.pem`.

Best regards.

**Comment by FernandoGarcia** [comment]
Pointing directly to `/etc/letsencrypt/archive/domain.com/chain4.pem` or `/etc/letsencrypt/archive/domain.com/fullchain4.pem` I can't see the error above but I can see this error:

```
2023-05-24T13:40:20.670647-03:00 [error] supervisor: 'esockd_connection_sup - <0.2298.0>', errorContext: connection_shutdown, reason: {ssl_error,{options,{cacertfile,[]}}}, offender: [{pid,<0.2874.0>},{name,connection},{mfargs,{emqx_connection,start_link,[#{enable_authn => true,limiter => undefined,listener => {ssl,default},zone => default}]}}]
```

For now I just want apply SSL to dashboard.

My settings are the same as in [documentation for V5](https://www.emqx.io/docs/en/v5/security/ssl.html).

```
listeners.ssl.default {
  bind = "0.0.0.0:8883"

  access_rules = [
    "allow all"
  ]

  proxy_protocol = false
  proxy_protocol_timeout = 3s

  ssl_options.keyfile = "etc/certs/key.pem"
  ssl_options.certfile = "etc/certs/cert.pem"
  ssl_options.cacertfile = "etc/certs/cacert.pem"

  # to verify client certs
  ssl_options.verify = verify_peer

  ssl_options.versions = ["tlsv1.3", "tlsv1.2", "tlsv1.1", "tlsv1"]
  tcp_options.backlog = 1024
  tcp_options.buffer = 4KB
}

```

**Reply by zmstone** [reply]
`{ssl_error,{options,{cacertfile,[]}}}` indicates the caceretfile is invalid.
maybe it's not a valid pem? or it could be an access issue for `emqx` user.

**Comment by FernandoGarcia** [comment]
There's some misunderstanding from my side. 

I'm looking in the wrong configuration. 

I have add listener to dashboard but I can find a full example in the [documentation](https://docs.emqx.com/en/enterprise/v5.0/configuration/configuration-manual.html#dashboard).

From I can find  it should look like [this](https://github.com/emqx/emqx/issues/3720#issue-697386838).

I tried:

```
dashboard.listeners.http.bind = 18083
dashboard.listeners.https.enable = true
dashboard.listeners.https.bind = 18084
dashboard.listeners.https.cacertfile = "/etc/letsencrypt/live/domain.com/chain.pem"
dashboard.listeners.https.certfile = "/etc/letsencrypt/live/domain.com/cert.pem"
dashboard.listeners.https.keyfile = "/etc/letsencrypt/live/domain.com/privkey.pem"
```

In this way there's no errors on log but the browser returns: ERR_CONNECTION_CLOSED

I'm sure that the 18084 port is not blocked by firewall because I can open the dashboard changing the HTTP listener to it.

Edit: 

I tried copy the certicates to /etc/emqx/certs and changed the owner to emqx:emqx but doesn't worked.

**Reply by lafirest** [reply]
It seems your configuration is right, could you please set the log level to `debug` to view what outputs when a client connects to the 18084.

**Reply by FernandoGarcia** [reply]
Hi!

Finally it worked.

There's permissions issues when path are from `/etc/letsencrypt/live/domain.com/`.

I had to copy certificates to `/etc/emqx/certs/` then change the ownership to `emqx:emqx` and set permissions to `644`.

I'll have problems with certificate renovation. There's some way to solve it without write a custom script?

Also I had the line below in the configuration and emqx was crashing on restart.

`dashboard.listeners.https.tls_versions = ["tlsv1.3", "tlsv1.2", "tlsv1.1", "tlsv1"] `

Also tried:

`dashboard.listeners.https.tls_versions = tlsv1.3,tlsv1.2,tlsv1.1,tlsv1`

The errors are like this:
```

2023-05-25T14:37:43.890445-03:00 [notice] Application: emqx. Exited: stopped. Type: permanent.
2023-05-25T14:37:43.890637-03:00 [debug] msg: stopping_app, mfa: emqx_machine_boot:stop_one_app/1, line: 80, app: cowboy
2023-05-25T14:37:43.895167-03:00 [notice] Application: cowboy. Exited: stopped. Type: permanent.
2023-05-25T14:37:43.895351-03:00 [debug] msg: stopping_app, mfa: emqx_machine_boot:stop_one_app/1, line: 80, app: ranch
2023-05-25T14:37:43.900577-03:00 [notice] Application: ranch. Exited: stopped. Type: permanent.
2023-05-25T14:37:43.900767-03:00 [debug] msg: stopping_app, mfa: emqx_machine_boot:stop_one_app/1, line: 80, app: esockd
2023-05-25T14:37:43.906225-03:00 [notice] Application: esockd. Exited: stopped. Type: permanent.
2023-05-25T14:37:43.906367-03:00 [debug] msg: stopping_app, mfa: emqx_machine_boot:stop_one_app/1, line: 80, app: gproc
2023-05-25T14:37:43.912334-03:00 [notice] Application: gproc. Exited: stopped. Type: permanent.
2023-05-25T14:37:43.912533-03:00 [debug] msg: stopping_app, mfa: emqx_machine_boot:stop_one_app/1, line: 80, app: emqx_conf
2023-05-25T14:37:43.915055-03:00 [notice] Application: emqx_conf. Exited: stopped. Type: temporary.
2023-05-25T14:37:43.951563-03:00 [notice] msg: Mria is stopped, mfa: mria_app:stop/1, line: 43
```

Best regards.

**Reply by zmstone** [reply]
the logs here are not quite helpful as they are just application shutdown/stop logs. could you send the full logs ?

**Comment by FernandoGarcia** [comment, marked as answer]
I did this script in case someone is interested.

```bash
#!/bin/sh
RSYNC_COMMAND=$(sudo rsync -uaL --progress /etc/letsencrypt/live/domain.com/ /etc/emqx/certs/)

    if [ $? -eq 0 ]; then
        if [ -n "${RSYNC_COMMAND}" ]; then
         find /etc/emqx/certs/ -name "*.pem" -exec sudo chown -R emqx:emqx {} \;
         find /etc/emqx/certs/ -name "*.pem" -exec sudo chmod 644 {} \;
        fi
    fi
```
It can be used as [post validation hook](https://eff-certbot.readthedocs.io/en/stable/using.html#pre-and-post-validation-hooks) for certbot.

**Reply by id** [reply]
@FernandoGarcia it should be possible to have owner and permissions set by rsync itself, see e.g. https://serverfault.com/questions/564385/rsync-command-issues-owner-and-group-permissions-doesn%c2%b4t-change/656494#656494

**Reply by FernandoGarcia** [reply]
Hi!

Thank you for the suggestion.

I tried this options but something is going wrong.

The files and permission seems applied when doing a `ls -la` but the browser shows an error in SSL handshake.

I tried this ways:

```shell
RSYNC_COMMAND=$(sudo rsync -uaLog --chown=emqx:emqx --chmod=644 --progress /etc/letsencrypt/live/domain.com/ /etc/emqx/certs/)

RSYNC_COMMAND=$(sudo rsync -uaL --chown=emqx:emqx --chmod=644 --progress /etc/letsencrypt/live/domain.com/ /etc/emqx/certs/)
```

Best regards.

**Reply by id** [reply]
Could you try to add `--copy-unsafe-links` flag?
```
--copy-unsafe-links
    This tells rsync to copy the referent of symbolic links that point outside the copied tree. Absolute symlinks are also treated like ordinary files, and so are any symlinks in the source path itself when --relative is used. This option has no additional effect if --copy-links was also specified. 
```

**Reply by FernandoGarcia** [reply]
Hi!
Thanks for your help.
Unfortunately with this option also doesn't works.
Best regards.

**Comment by zmstone** [comment]
Here is a tutorial which explains in detail what the files are and how to configure them.
It's based on v4, but the three files are the same.
https://www.youtube.com/watch?v=W7SedpZzQdo

**Reply by FernandoGarcia** [reply]
Hi!
Thank you for your answer.
The problem was solved. 
The problem is related to permissions.
The solution is above.
Best regards.

**Comment by zmstone** [comment]
Hi @FernandoGarcia 
Thank you for getting back.
I have changed this to a Q&A and marked your reply as the answer.
